src/main/java/com/scalar/db/storage/jdbc/JDBCService.java

```diff
@@ -0,0 +1,213 @@
+      throws SQLException {
+    addProjectionsForKeys(get);
+    Table table = getTable(get, defaultSchema, defaultTable);
+    @Nullable Key clusteringKey = get.getClusteringKey().orElse(null);
+
+    operationChecker.checkGet(table, get.getProjections(), get.getPartitionKey(), clusteringKey);
+
+    SelectQuery selectQuery =
+        queryBuilder
+            .select(get.getProjections())
+            .from(table)
+            .where(get.getPartitionKey(), clusteringKey)
+            .build();
+
+    PreparedStatement preparedStatement = selectQuery.prepareAndBind(connection);
+    try (ResultSet resultSet = preparedStatement.executeQuery()) {
+      if (resultSet.next()) {
+        return Optional.of(selectQuery.getResult(resultSet));
+      }
+      return Optional.empty();
+    }
+  }
+
+  public Scanner scan(
+      Scan scan,
+      Connection connection,
+      @Nullable String defaultSchema,
+      @Nullable String defaultTable)
+      throws SQLException {
+    addProjectionsForKeys(scan);
+    Table table = getTable(scan, defaultSchema, defaultTable);
+
+    @Nullable Key startClusteringKey = scan.getStartClusteringKey().orElse(null);
+    @Nullable Key endClusteringKey = scan.getEndClusteringKey().orElse(null);
+
+    operationChecker.checkScan(
+        table,
+        scan.getProjections(),
+        scan.getPartitionKey(),
+        startClusteringKey,
+        endClusteringKey,
+        scan.getLimit(),
+        scan.getOrderings());
+
+    SelectQuery selectQuery =
+        queryBuilder
+            .select(scan.getProjections())
+            .from(table)
+            .where(
+                scan.getPartitionKey(),
+                startClusteringKey,
+                scan.getStartInclusive(),
+                endClusteringKey,
+                scan.getEndInclusive())
+            .orderBy(scan.getOrderings())
+            .limit(scan.getLimit())
+            .build();
+
+    PreparedStatement preparedStatement = selectQuery.prepareAndBind(connection);
+    ResultSet resultSet = preparedStatement.executeQuery();
+    return new ScannerImpl(selectQuery, connection, resultSet);
+  }
+
+  private void addProjectionsForKeys(Selection selection) {
+    if (selection.getProjections().size() == 0) { // meaning projecting all
+      return;
+    }
+    selection.getPartitionKey().forEach(v -> selection.withProjection(v.getName()));
+    selection
+        .getClusteringKey()
+        .ifPresent(k -> k.forEach(v -> selection.withProjection(v.getName())));
+  }
+
+  public boolean put(
+      Put put, Connection connection, @Nullable String defaultSchema, @Nullable String defaultTable)
+      throws SQLException {
+    Table table = getTable(put, defaultSchema, defaultTable);
+    @Nullable Key clusteringKey = put.getClusteringKey().orElse(null);
+    @Nullable MutationCondition condition = put.getCondition().orElse(null);
+
+    operationChecker.checkPut(
```

Comment:
How to check seems checker's responsibility so I think it would be better it is determined and implemented by the check method. By doing this, the users of the checker don't need to care what to pass to check any more.

```suggestion
    operationChecker.check(table, delete);
```